### ISSUE TITLE
Undefined variable rule does not understand anonymous classes

### ISSUE THREAD

**Issue by ndench** [phpmd/phpmd#718]
- PHPMD version: 2.8.1
- PHP Version: 7.3
- Installation type: composer
- Operating System / Distribution & Version: Ubuntu 16.04

### Current Behavior
The undefined variable rule does not pick up parameters passed into a method of an anonymous class.

```php
final class FooChecker
{
    public function getChecker()
    {
        return new class() {
            public function checkFoo(User $user): bool
            {
                // Avoid using undefined variables such as $user
                return $user->addFailedLoginAttempt() > 1;
            }
        };
    }
}
```

### Expected Behavior
`$user` is defined in the above example, so it should not produce an undefined variable error.

### Steps To Reproduce:
Run the undefined variable rule over the above example

## Checks before submitting
* [x] Be sure that there isn't already an issue about this. See: [Issues list](https://github.com/phpmd/phpmd/issues)
* [x] Be sure that there isn't already a pull request about this. See: [Pull requests](https://github.com/phpmd/phpmd/pulls)
* [x] I have added every step to reproduce the bug.
* [x] If possible I added relevant code examples.
* [x] This issue is about 1 bug and nothing more.
* [x] The issue has a descriptive title. For example:  "JSON rendering failed on Windows for filenames with space".

**Comment by kylekatarnls** [comment]
I was analyzing a similar problem with anonymous classes, finding that methods was parsed as FunctionNode instead of MethodNode. And this triggered `undefined` warning for `$this` I think it may be the same problem here.

**Comment by yutv** [comment]
The same error in `foreach ($array as &$item)` but works (no error) in `foreach ($array as $item)`. It should work in both cases w/o error since the `$item` variable is defined.

**Example 1:**

Foo.php

    <?php

    class Foo
    {
        public function bar(array $array)
        {
            foreach ($array as &$item) {
                $item['value']++;
            }
            return $array;
        }
    }

**Result:**

    $ vendor/bin/phpmd Foo.php text cleancode && echo "Success"
    /home/user/project/Foo.php:7  Avoid using undefined variables such as '$item' which will lead to PHP notices.
    /home/user/project/Foo.php:8  Avoid using undefined variables such as '$item' which will lead to PHP notices.

**Example 2:**

Foo.php

    <?php

    class Foo
    {
        public function bar(array $array)
        {
            foreach ($array as $item) {
                $item['value']++;
            }
            return $array;
        }
    }

    $ vendor/bin/phpmd Foo.php text cleancode && echo "Success"
    Success

----
Affected versions: 2.8.1, 2.8.0, 2.8.0-beta.2, 2.8.0-beta.1
No issue in: 2.7.0

**Comment by kylekatarnls** [comment]
Hi @yutv there is no error in 2.7.0 because "undefined variable" rule did not exist before. Your error has not the same root cause. `as &$item` is saw as a reading instead as a writing (as it is for `as $item`) which explain you have the error on both lines 7 and 8.

This could deserve to open an other issue as it will need a completely different fix.

**Comment by yutv** [comment]
Created separate Issue https://github.com/phpmd/phpmd/issues/720.

**Comment by kylekatarnls** [comment]
Thanks.

**Comment by kylekatarnls** [comment]
This will be fixed in 2.9.0. You can try the fix requiring `"phpmd/phpmd": "dev-master"`
